pkg/service/replicator/replicate.go

```diff
@@ -0,0 +1,218 @@
+	uploadService client.Connection,
+) (*Service, error) {
+
+	repl := &Service{
+		uploadService: uploadService,
+		id:            signer,
+		capabilities:  caps,
+		receipts:      rctStore,
+	}
+
+	replicationQueue := jobqueue.NewJobQueue[*Task](
+		jobqueue.JobHandler(repl.replicate),
+		jobqueue.WithErrorHandler(func(err error) {
+			log.Errorf("error while handling replication request: %s", err)
+		}),
+	)
+
+	repl.queue = replicationQueue
+
+	return repl, nil
+
+}
+
+func (r *Service) Replicate(ctx context.Context, task *Task) error {
+	return r.queue.Queue(ctx, task)
+}
+
+func (r *Service) Start() error {
+	r.queue.Startup()
+	return nil
+}
+
+func (r *Service) Stop(ctx context.Context) error {
+	return r.queue.Shutdown(ctx)
+}
+
+func (r *Service) replicate(ctx context.Context, task *Task) error {
+	// pull the data from the source
+	replicaResp, err := http.Get(task.Source.String())
+	if err != nil {
+		return fmt.Errorf("http get replication source (%s) failed: %w", task.Source.String(), err)
+	}
+
+	// stream the source to the sink
+	req, err := http.NewRequest(http.MethodPut, task.Sink.String(), replicaResp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to create replication sink request: %w", err)
+	}
+	req.Header = replicaResp.Header
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf(
+			"failed http PUT to replicate blob %s from %s to %s failed: %w",
+			task.Blob.Digest,
+			task.Source.String(),
+			task.Sink.String(),
+			err,
+		)
+	}
+	// verify status codes
+	if res.StatusCode >= 300 || res.StatusCode < 200 {
+		topErr := fmt.Errorf(
+			"unsuccessful http PUT to replicate blob %s from %s to %s status code %d",
+			task.Blob.Digest,
+			task.Source.String(),
+			task.Sink.String(),
+			res.StatusCode,
+		)
+		resData, err := io.ReadAll(res.Body)
+		if err != nil {
+			return fmt.Errorf("%s failed to read replication sink response body: %w", topErr, err)
+		}
+		return fmt.Errorf("%s response body: %s: %w", topErr, resData, err)
+	}
+
+	acceptResp, err := r.capabilities.BlobAccept(ctx, &capabilities.BlobAcceptRequest{
+		Space: task.Space,
+		Blob:  task.Blob,
+		Put: blob.Promise{
+			UcanAwait: blob.Await{
+				Selector: ".out.ok",
+				// TODO IDK what this does, or what to put here, or if we even need it?
+				Link: task.Invocation.Link(),
+			},
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to accept replication source blob %s: %w", task.Blob.Digest, err)
+	}
+
+	var forks []fx.Effect
+	forks = append(forks, fx.FromInvocation(acceptResp.Claim))
+
+	var pdpLink *ucan.Link
+	if acceptResp.Piece != nil {
+		// generate the invocation that will complete when aggregation is complete and the piece is accepted
+		pieceAccept, err := cap_pdp.Accept.Invoke(
+			r.id,
+			// TODO validate this is the correct audience
+			r.uploadService.ID().DID(),
+			r.id.DID().String(),
+			cap_pdp.AcceptCaveats{
+				Piece: *acceptResp.Piece,
+			}, delegation.WithNoExpiration())
```

Comment:
to reviewer: need to validate I have the issuer, audience, and resource defined correctly here.

Reply:
no should be:
issuer: the storage node
audience: the storage node
resource: the storage node

see handler in `pkg/service/storage/ucan.go`

However, why is this not itself in prob replicate? why not have acceptResp.Piece be an optional link to an invocation, instead of the Piece itself?